lib/src/main/scala/spinal/lib/com/jtag/sim/JtagVpi.scala

```diff
@@ -0,0 +1,158 @@
+      else {
+        inputStream.readFully(buffer)
+
+        val vpiCmd = deserializeVpiCmd(buffer)
+
+        vpiCmd match {
+          case VpiCmd(Cmds.RESET, _, _, _, _) => {
+            driver.doResetTap()
+          }
+          case VpiCmd(Cmds.TMS_SEQ, bufferOut, _, _, nbBits) => {
+            val tmsSeq = bufferOut.map(x => (0 until 8).map(i => (x & (1 << i)) != 0)).flatten.take(nbBits)
+            driver.doTmsSeq(tmsSeq)
+          }
+          case VpiCmd(Cmds.SCAN_CHAIN, bufferOut, bufferIn, length, nbBits) => {
+            // bufferOut is a byte array, we need to convert it to a boolean array
+            // bufferIn is a byte array, we need to convert it to a boolean array
+            // each byte of bufferOut is a sequence of 8 bits (which is why nbBits exists)
+            // convert bufferOut to a boolean array
+            val tdiSeq = bufferOut.map(x => (0 until 8).map(i => (x & (1 << i)) != 0)).flatten.take(nbBits)
+            val tdoSeq = driver.doScanChain(tdiSeq, false)
+            // convert tdoSeq to a byte array grouped by 8 bits
+            val tdoSeqBytes = tdoSeq.grouped(8).map(_.foldLeft(0)((acc, b) => (acc << 1) | (if (b) 1 else 0))).toArray
+            // copy the result to bufferIn
+            for (i <- 0 until length) {
+              bufferIn(i) = tdoSeqBytes(i).toByte
+            }
+            // create packet to send to the client, exactly the same as the received packet but with the tdoSeq
+            val vpiCmd = VpiCmd(Cmds.SCAN_CHAIN, bufferOut, bufferIn, length, nbBits)
+            serialize(buffer, vpiCmd)
+            outputStream.write(buffer)
+          }
+          case VpiCmd(Cmds.SCAN_CHAIN_FLIP_TMS, bufferOut, bufferIn, length, nbBits) => {
+            // same as SCAN_CHAIN but with the last TMS set to 1
+            val tdiSeq = bufferOut.map(x => (0 until 8).map(i => (x & (1 << i)) != 0)).flatten.take(nbBits)
+            val tdoSeq = driver.doScanChain(tdiSeq, true)
+            val tdoSeqBytes =
+              tdoSeq.grouped(8).map(_.reverse.foldLeft(0)((acc, b) => (acc << 1) | (if (b) 1 else 0))).toArray
+            for (i <- 0 until length) {
+              bufferIn(i) = tdoSeqBytes(i).toByte
+            }
+            val vpiCmd = VpiCmd(Cmds.SCAN_CHAIN_FLIP_TMS, bufferOut, bufferIn, length, nbBits)
+            serialize(buffer, vpiCmd)
+            outputStream.write(buffer)
+          }
+          case VpiCmd(Cmds.STOP_SIMU, _, _, _, _) => {
+            println("Stop simulation")
+            simSuccess()
+          }
+        }
+      }
+    }
+  }
+
+  def deserializeVpiCmd(buffer: Array[Byte]): VpiCmd = {
+    val byteBuffer = ByteBuffer.wrap(buffer)
+    byteBuffer.order(ByteOrder.LITTLE_ENDIAN)
+    val cmd = byteBuffer.getInt match {
+      case 0 => Cmds.RESET
+      case 1 => Cmds.TMS_SEQ
+      case 2 => Cmds.SCAN_CHAIN
+      case 3 => Cmds.SCAN_CHAIN_FLIP_TMS
+      case 4 => Cmds.STOP_SIMU
+    }
+    val bufferOut = new Array[Byte](XFERT_MAX_SIZE)
```

Comment:
since everything is synchronous you could stick with a single set of buffers?

Reply:
I don't think it helps anything since both buffers (in/out tdi/o scan chain) are sent on receive and send of the vpi packet...

Reply:
my thought was to remove a few allocations/frees, but in the grand scheme of things the one where you do the byte -> boolean conversion is much heavier...
Let's keep it and maybe have some optimization potential ;-)